hub/core/lowlevel.py

```diff
@@ -0,0 +1,272 @@
+
+    def __setitem__(self, idx: int, byte: int) -> None:
+        self._c_array[idx] = byte
+
+    def __getitem__(self, idx: Union[int, slice]) -> Union[int, "Pointer"]:
+        if isinstance(idx, int):
+            return self._c_array[idx]
+        elif isinstance(idx, slice):
+            assert idx.step is None
+            start = idx.start
+            end = idx.stop
+            n = self.size
+            if start is None:
+                start = 0
+            elif start < 0:
+                start += n
+            if end is None:
+                end = n
+            elif end < 0:
+                end += n
+            assert start >= 0 and start < n
+            assert end >= start and end <= n
+            ret = Pointer(self.address + start, end - start)
+            ret._refs.append(self._c_array)
+            return ret
+
+    @property
+    def memoryview(self):
+        return memoryview(self._c_array)
+
+    @property
+    def bytes(self):
+        return bytes(self._c_array)
+
+    @property
+    def bytearray(self):
+        return bytearray(self._c_array)
+
+    def __len__(self):
+        return self.size
+
+
+def malloc(size: int) -> Pointer:
+    return Pointer(c_array=(ctypes.c_byte * size)())
+
+
+def memcpy(dest: Pointer, src: Pointer, count=None) -> None:
+    if count is None:
+        count = src.size
+    ctypes.memmove(dest.address, src.address, count)
+
+
+def _write_pybytes(ptr: Pointer, byts: Union[bytes, memoryview]) -> Pointer:
+    memcpy(ptr, _ndarray_to_ptr(np.frombuffer(byts, dtype=np.byte)))
+    return ptr + len(byts)
+
+
+def _ndarray_to_ptr(arr: np.ndarray) -> Pointer:
+    return Pointer(arr.__array_interface__["data"][0], arr.itemsize * arr.size)
+
+
+def _pybytes_to_c_array(byts: bytes) -> Pointer:
+    return Pointer(
+        np.frombuffer(byts, dtype=np.byte).__array_interface__["data"][0], len(byts)
+    )
+
+
+def _infer_chunk_num_bytes(
+    version: str,
+    shape_info: np.ndarray,
+    byte_positions: np.ndarray,
+    data: Optional[Union[Sequence[bytes], Sequence[memoryview]]] = None,
+    len_data: Optional[int] = None,
+) -> int:
+    # NOTE: Assumption: version string contains ascii characters only (ord(c) < 128)
+    # NOTE: Assumption: len(version) < 256
+    assert len(version) < 256
+    assert max((map(ord, version))) < 128
+    # assert shape_info.ndim == 2
+    # assert byte_positions.ndim == 2
+    # version_slice_size = 1 + len(version)
+    # shape_info_slice_size = 4 + 4 + shape_info.nbytes
+    # byte_positions_slice_size = 4 + byte_positions.nbytes
+    # data_slice_size = sum(map(len, data))
+    if len_data is None:
+        len_data = sum(map(len, data))  # type: ignore
+    return len(version) + shape_info.nbytes + byte_positions.nbytes + len_data + 13
+
+
+def encode_chunk(
+    version: str,
+    shape_info: np.ndarray,
+    byte_positions: np.ndarray,
+    data: Union[Sequence[bytes], Sequence[memoryview]],
+    len_data: Optional[int] = None,
+) -> memoryview:
```

Comment:
can we move encoders into a new file? ideally we should delete the `lowlevel` file

Reply:
i recommend putting `encode_chunk` in a new module: `hub.core.meta.encode.chunk`